Sources/SwiftProtobuf/Message+AsyncMessageSequence.swift

```diff
@@ -0,0 +1,232 @@
+  ///   - options: The BinaryDecodingOptions to use.
+  /// - Returns: An asynchronous sequence of messages read from the `AsyncSequence` of bytes.
+  /// - Throws: `BinaryDecodingError` if decoding fails, throws
+  ///           `BinaryDelimited.Error` for some reading errors,
+  ///           `BinaryDecodingError.truncated` if the stream ends before fully decoding a
+  ///           message or a delimiter,
+  ///           `BinaryDecodingError.malformedProtobuf`if a delimiter could not be read and
+  ///           `BinaryDecodingError.tooLarge` if a size delimiter of 2GB or greater is found.
+  public init(
+    baseSequence: Base,
+    extensions: ExtensionMap? = nil,
+    partial: Bool = false,
+    options: BinaryDecodingOptions = BinaryDecodingOptions()
+  ) {
+    self.baseSequence = baseSequence
+    self.extensions = extensions
+    self.partial = partial
+    self.options = options
+  }
+  
+  /// An asynchronous iterator that produces the messages of this asynchronous sequence
+  @frozen public struct AsyncIterator: AsyncIteratorProtocol, Sendable
+  where Base.AsyncIterator: Sendable {
+    
+    private var iterator: Base.AsyncIterator?
+    private let extensions: ExtensionMap?
+    private let partial: Bool
+    private let options: BinaryDecodingOptions
+    
+    init(
+      iterator: Base.AsyncIterator,
+      extensions: ExtensionMap?,
+      partial: Bool,
+      options: BinaryDecodingOptions
+    ) {
+      self.iterator = iterator
+      self.extensions = extensions
+      self.partial = partial
+      self.options = options
+    }
+    
+    /// Aysnchronously reads the next varint
+    private mutating func nextVarInt() async throws -> UInt64? {
+      var messageSize: UInt64 = 0
+      var shift: UInt64 = 0
+      
+      while let byte = try await iterator?.next() {
+        messageSize |= UInt64(byte & 0x7f) << shift
+        shift += UInt64(7)
+        if shift > 35 {
+          throw BinaryDecodingError.malformedProtobuf
+        }
+        if (byte & 0x80 == 0) {
+          return messageSize
+        }
+      }
+      if (shift > 0) {
+        // The stream has ended inside a varint.
+        throw BinaryDecodingError.truncated
+      }
+      return nil // End of stream reached.
+    }
+    
+    /// Asynchronously advances to the next message and returns it, or ends the
+    /// sequence if there is no next message.
+    ///
+    /// - Returns: The next message, if it exists, or `nil` to signal the end of
+    ///   the sequence.
+    public mutating func next() async throws -> M? {
+      guard let messageSize = try await nextVarInt() else {
+        iterator = nil
+        return nil
+      }
+      if messageSize == 0 {
+        return try M(
+          serializedBytes: [],
+          extensions: extensions,
+          partial: partial,
+          options: options
+        )
+      } else if messageSize > 0x7fffffff {
+        throw BinaryDecodingError.tooLarge
```

Comment:
We should `nil` out the iterator here

Reply:
Done. I also `nil`ed the iterator in the early returns in the `nextVarInt` function.